### ISSUE TITLE
support for proxy repos

### ISSUE THREAD

**Issue by robey** [sbt/sbt#414]
our builds-for-deploy always take place on a build machine without internet access. instead of a set of maven repos, only one is listed: an internal "proxy" repo, running artifactory.

currently, we have to use a bunch of boilerplate in both plugins.sbt and our "sbt-package-dist" plugin to allow replacing the default repo list with a proxy repo in this case.

it would be great if sbt natively supported proxy repos. something from the environment -- an environment variable or property or something -- would tell sbt that a proxy repo should replace any repo list. we use environment variable "SBT_PROXY_REPO".

the implementation of this feature is most glaring in plugins.sbt, which has to be copied around from project to project as boilerplate, so even if this feature was only implemented for plugin resolution, that would still be a win.

here's our current plugins.sbt boilerplate:

```
sbtResolver <<= (sbtResolver) { r =>
  Option(System.getenv("SBT_PROXY_REPO")) map { x =>
    Resolver.url("proxy repo for sbt", url(x))(Resolver.ivyStylePatterns)                                             
  } getOrElse r
}

resolvers <<= (resolvers) { r =>
  (Option(System.getenv("SBT_PROXY_REPO")) map { url =>
    Seq("proxy-repo" at url)
  } getOrElse {
    r ++ Seq(
      "twitter.com" at "http://maven.twttr.com/",
      "scala-tools" at "http://scala-tools.org/repo-releases/",
      "maven" at "http://repo1.maven.org/maven2/",
      "freemarker" at "http://freemarker.sourceforge.net/maven2/"
    )
  }) ++ Seq("local" at ("file:" + System.getProperty("user.home") + "/.m2/repo/"))
}

externalResolvers <<= (resolvers) map identity
```

**Comment by harrah** [comment]
Yes, the idea sounds good to me.  System property and/or environment variable are fine.

**Comment by jsuereth** [comment]
It's not quite good enough to have one proxy.   If you're proxying both Ivy + Maven, you need to set up the same URL as both a maven repo and an ivy repo.   This is, of course, only possible using Artifactory, since Nexus can't handle proxying non-maven repos AFAICT.

**Comment by robey** [comment]
ah, yeah, we're using artifactory. maybe the ivy one should be called SBT_PROXY_REPO_IVY? you could set them both to the same url if you use artifactory.

(it would simplify things a lot if sbt published in maven-style, instead of ivy-style. it's the last project i know of that's still fighting this battle. ... i'm no lover of maven, but i think the war was lost a while ago.)

**Comment by jsuereth** [comment]
I just use the same URL with ivy-style paths and again as a maven repo.   It's two resolvers, but one repo.  I also hook it into sbt.boot.properties so all SBT resolves that way.

**Comment by harrah** [comment]
Fixed by #472
